sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -76,24 +76,34 @@ object ExprCode {
 /**
  * State used for subexpression elimination.
  *
- * @param isNull A term that holds a boolean value representing whether the expression evaluated
- *               to null.
- * @param value A term for a value of a common sub-expression. Not valid if `isNull`
- *              is set to `true`.
+ * @param eval The source code for evaluating the subexpression.
+ * @param children The sequence of subexpressions as the children expressions. Before
+ *                 evaluating this subexpression, we should evaluate all children
+ *                 subexpressions first. This is used if we want to selectively evaluate
+ *                 particular subexpressions, instead of all at once. In the case, we need
+ *                 to make sure we evaluate all children subexpressions too.
  */
-case class SubExprEliminationState(isNull: ExprValue, value: ExprValue)
+case class SubExprEliminationState(
+  eval: ExprCode,
+  children: Seq[SubExprEliminationState])
+
+object SubExprEliminationState {
+  def apply(
+      eval: ExprCode,
+      children: Seq[SubExprEliminationState] = Seq.empty): SubExprEliminationState = {
```

Comment:
nit: `def apply(eval: ExprCode): ...`. If `children` parameter is also provided, the default case class apply should work.

Reply:
This is for @maropu's comment https://github.com/apache/spark/pull/32980#discussion_r659400977.

Reply:
Or you mean to also add `def apply(eval: ExprCode)` here?

Reply:
Added `def apply(eval: ExprCode)`.